keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+// while handling delayed allocation.
+type DriverSchedulingStatus struct {
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod. A resource driver can
+	// set this immediately when a ResourceClaim gets created and, for
+	// example, provide a static selector that uses labels.
+	//
+	// Setting this field is optional. If nil, all nodes are candidates.
+	SuitableNodes *core.NodeSelector
+
+	// A change of the PotentialNodes field triggers a check in the driver
+	// on which of those nodes the resource might be made available. It
+	// then excludes nodes by listing those where that is not the case in
+	// UnsuitableNodes.
+	//
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// This can change, so the driver must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	UnsuitableNodes []string
+}
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// Attributes contains arbitrary data returned by the driver after a
+	// successful allocation.  This data is passed to the driver for all
+	// operations involving the allocated resource. This is opaque for
+	// Kubernetes.  Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// The attributes must be sufficient to deallocate the resource because
+	// the ResourceClass might not be available anymore when deallocation
+	// starts.
+	Attributes map[string]string
+
+	// This field will get set by the resource driver after it has
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// A resource driver may already set this before the resource is
+	// allocated. The scheduler will then check this field in addition to
+	// UnsuitableNodes to filter out nodes where the resource cannot be
+	// allocated.
+	//
+	// Setting this field is optional. If nil, the resource is available
+	// everywhere.
+	AvailableOnNodes *core.NodeSelector
+
+	// SharedResource determines whether the resource supports more
+	// than one user at a time.
+	SharedResource bool
+}
+
+type PodSpec {
+   ...
+   // ResourceClaims defines which ResourceClaims must be allocated
+   // and reserved before the Pod is allowed to start. The resources
+   // will be made available to those containers which reference them
+   // by name.
+   ResourceClaims []PodResourceClaim
+   ...
+}
+
+type  ResourceRequirements {
+   Limits ResourceList
+   Requests ResourceList
+   ...
+   // The entries are the names of resources in PodSpec.ResourceClaims
+   // that are used by the container.
+   Claims []string
```

Comment:
By analogy - volumes have "mount points".  Do we need the same?  How does a container know where to find the devices it claimed?  E.g. major/minor numbers or device nodes or FS paths or IP addresses or .... ?

Do we end up with a parameter block for each container-claim?   Do we need a downward API to expose this info?  Or are pods left to go read ResourceClaims themselves?

Reply:
> By analogy - volumes have "mount points". Do we need the same? How does a container know where to find the devices it claimed? E.g. major/minor numbers or device nodes or FS paths or IP addresses or .... ?

We discussed this among the people defining CDI and the conclusion there was that we don't need it. Processes find their resources in a way that is defined by the resource driver vendor, for example via env variables.

Treating all containers alike simplifies the implementation (one gRPC call from kubelet per node  instead of one per node and then one for each container). Let's go with that until we have stronger evidence that additional flexibility is needed. This might come from some resource driver vendor who finds the current model too limiting.

But as this might happen, should we already plan for it and make this `Claims` a slice of structs where currently the only entry is the name?

Reply:
> Processes find their resources in a way that is defined by the resource driver vendor, for example via env variables.

How does a driver vendor inject env variables into a Pod?

If I am a user using a ResourceClass, I am ready 2 indirections away from knowing anything about the driver.  It's a little "devil's advocate", but if I have to know which driver is providing the device, then what value does ResourceClass provide?  I could just make my claim reference a driver-specific CR and skip one indirection.

> But as this might happen, should we already plan for it and make this Claims a slice of structs where currently the only entry is the name?

I guess there are 2 parts to this - some things the user can ask for (e.g. mount point) and some things the driver has to provide (e.g. dynamic major/minor values).  I worry that "things the user can ask for" is yet another arbitrary Claim->spec->params CR and that "things the driver has to provide" is yet another arbitrary Claim->status->params CR.  

How can we avoid that?

Reply:
> How does a driver vendor inject env variables into a Pod?

Through CDI. It's one of the things it can do.

> if I have to know which driver is providing the device, then what value does ResourceClass provide

It provides a safe way for the cluster admin to add additional parameters. 

> I worry that "things the user can ask for" is yet another arbitrary Claim->spec->params CR and that "things the driver has to provide" is yet another arbitrary Claim->status->params CR.

I agree.

> How can we avoid that?

By not supporting it? Seriously, it does add lots of complications and there is no evidence yet that it is needed.

Reply:
> Through CDI. It's one of the things it can do.

Well that's interesting - I didn't see it called out here.   Ahh, it's in the CDI spec.  In general we want users to tell us which variables to us, not the other way around.  I'll have to think about this more

> By not supporting it?

+1 until we know more.  If you want to make this a list of struct, I would not object

Reply:
Let's stick with string because a struct with a single string is unnecessarily complex at the moment.

I added an UNRESOLVED to capture that this may have to be reconsidered.